README.md
# data-validation
Efficient and Easy Data Validation with [Cats](https://github.com/typelevel/cats)

Video and slides from Typelevel Summit NYC 2017
* Slides: https://speakerdeck.com/danielasfregola/easy-and-efficient-data-validation-with-cats-typelevel-summit-2017-nyc
* Video: https://www.youtube.com/watch?v=OkTfcyFohS0


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0] - 2020-05-27
### Changed
- Upgrade to Cats 2.1.0
- Switch from `ValidatedNel` to `ValidatedNec` which uses `Chain` as its base data structure. `Chain` allows constant time prepending and appending. This makes it especially efficient when used as a Monoid, e.g. with Validated or Writer.

## [1.0] - 2016-10-27
### Added
- Initial release

[Unreleased]: https://github.com/DanielaSfregola/data-validation
[2.0]: https://github.com/DanielaSfregola/data-validation/releases/v2.0
[1.0]: https://github.com/DanielaSfregola/data-validation/releases/v1.0


tools/codegen.scala
// Generates Accumulate function for several instances of ValidatedNel
// see src/main/scala/app/validation/AccumulateArities.scala

def validations(i: Int) = (1 to i).map(n => s"v$n: Validation[T$n]").mkString(",")

def vals(i: Int) = (1 to i).map(n => s"v$n").mkString(",")

def fntype(i: Int) = s"(${types(i)}) => Z"

def types(i: Int) = (1 to i).map(n => s"T$n").mkString(",")

def accumulatefn(i: Int) = s"""
  /** Accumulate function for Validation[T] of arity ${i} */
  def accumulate[${types(i)},Z](${validations(i)})(f: ${fntype(i)}): Validation[Z] =
    Apply[Validation].map$i(${vals(i)})(f)
"""

println(
  args(0) match {
    case "Accumulate" =>

      s"""/**
Defines an accumulate function for several instances of ValidatedNel[E, _]

Autogenerated using tools/codegen.sh
*/
package app.validation

import cats.Apply

trait AccumulateArities {
  ${2 to 22 map accumulatefn mkString ""}
}
 """
  }
)


src/main/scala/com/danielasfregola/datavalidation/CombineArities.scala
/**
Defines a combine function for several instances of ValidatedNel[E, _]

Autogenerated using tools/codegen.sh
*/
package com.danielasfregola.datavalidation

import cats.Apply
import cats.data.ValidatedNel

trait CombineArities[E]{

  private type Validation[T] = ValidatedNel[E, T]

  /** Combine function for Validation[T] of arity 2 */
  def combine[T1,T2,Z](v1: Validation[T1],v2: Validation[T2])(f: (T1,T2) => Z): Validation[Z] =
    Apply[Validation].map2(v1,v2)(f)

  /** Combine function for Validation[T] of arity 3 */
  def combine[T1,T2,T3,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3])(f: (T1,T2,T3) => Z): Validation[Z] =
    Apply[Validation].map3(v1,v2,v3)(f)

  /** Combine function for Validation[T] of arity 4 */
  def combine[T1,T2,T3,T4,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4])(f: (T1,T2,T3,T4) => Z): Validation[Z] =
    Apply[Validation].map4(v1,v2,v3,v4)(f)

  /** Combine function for Validation[T] of arity 5 */
  def combine[T1,T2,T3,T4,T5,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5])(f: (T1,T2,T3,T4,T5) => Z): Validation[Z] =
    Apply[Validation].map5(v1,v2,v3,v4,v5)(f)

  /** Combine function for Validation[T] of arity 6 */
  def combine[T1,T2,T3,T4,T5,T6,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6])(f: (T1,T2,T3,T4,T5,T6) => Z): Validation[Z] =
    Apply[Validation].map6(v1,v2,v3,v4,v5,v6)(f)

  /** Combine function for Validation[T] of arity 7 */
  def combine[T1,T2,T3,T4,T5,T6,T7,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7])(f: (T1,T2,T3,T4,T5,T6,T7) => Z): Validation[Z] =
    Apply[Validation].map7(v1,v2,v3,v4,v5,v6,v7)(f)

  /** Combine function for Validation[T] of arity 8 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8])(f: (T1,T2,T3,T4,T5,T6,T7,T8) => Z): Validation[Z] =
    Apply[Validation].map8(v1,v2,v3,v4,v5,v6,v7,v8)(f)

  /** Combine function for Validation[T] of arity 9 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9) => Z): Validation[Z] =
    Apply[Validation].map9(v1,v2,v3,v4,v5,v6,v7,v8,v9)(f)

  /** Combine function for Validation[T] of arity 10 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10) => Z): Validation[Z] =
    Apply[Validation].map10(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10)(f)

  /** Combine function for Validation[T] of arity 11 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11) => Z): Validation[Z] =
    Apply[Validation].map11(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11)(f)

  /** Combine function for Validation[T] of arity 12 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12) => Z): Validation[Z] =
    Apply[Validation].map12(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12)(f)

  /** Combine function for Validation[T] of arity 13 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13) => Z): Validation[Z] =
    Apply[Validation].map13(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13)(f)

  /** Combine function for Validation[T] of arity 14 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14) => Z): Validation[Z] =
    Apply[Validation].map14(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14)(f)

  /** Combine function for Validation[T] of arity 15 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15) => Z): Validation[Z] =
    Apply[Validation].map15(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15)(f)

  /** Combine function for Validation[T] of arity 16 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16) => Z): Validation[Z] =
    Apply[Validation].map16(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16)(f)

  /** Combine function for Validation[T] of arity 17 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16],v17: Validation[T17])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17) => Z): Validation[Z] =
    Apply[Validation].map17(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16,v17)(f)

  /** Combine function for Validation[T] of arity 18 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16],v17: Validation[T17],v18: Validation[T18])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18) => Z): Validation[Z] =
    Apply[Validation].map18(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16,v17,v18)(f)

  /** Combine function for Validation[T] of arity 19 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16],v17: Validation[T17],v18: Validation[T18],v19: Validation[T19])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19) => Z): Validation[Z] =
    Apply[Validation].map19(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16,v17,v18,v19)(f)

  /** Combine function for Validation[T] of arity 20 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16],v17: Validation[T17],v18: Validation[T18],v19: Validation[T19],v20: Validation[T20])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20) => Z): Validation[Z] =
    Apply[Validation].map20(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16,v17,v18,v19,v20)(f)

  /** Combine function for Validation[T] of arity 21 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20,T21,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16],v17: Validation[T17],v18: Validation[T18],v19: Validation[T19],v20: Validation[T20],v21: Validation[T21])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20,T21) => Z): Validation[Z] =
    Apply[Validation].map21(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16,v17,v18,v19,v20,v21)(f)

  /** Combine function for Validation[T] of arity 22 */
  def combine[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20,T21,T22,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16],v17: Validation[T17],v18: Validation[T18],v19: Validation[T19],v20: Validation[T20],v21: Validation[T21],v22: Validation[T22])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20,T21,T22) => Z): Validation[Z] =
    Apply[Validation].map22(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16,v17,v18,v19,v20,v21,v22)(f)

}
 


src/main/scala/app/validation/AccumulateArities.scala
/**
Defines an accumulate function for several instances of ValidatedNel[E, _]

Autogenerated using tools/codegen.sh
*/
package app.validation

import cats.Apply

trait AccumulateArities {
  
  /** Accumulate function for Validation[T] of arity 2 */
  def accumulate[T1,T2,Z](v1: Validation[T1],v2: Validation[T2])(f: (T1,T2) => Z): Validation[Z] =
    Apply[Validation].map2(v1,v2)(f)

  /** Accumulate function for Validation[T] of arity 3 */
  def accumulate[T1,T2,T3,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3])(f: (T1,T2,T3) => Z): Validation[Z] =
    Apply[Validation].map3(v1,v2,v3)(f)

  /** Accumulate function for Validation[T] of arity 4 */
  def accumulate[T1,T2,T3,T4,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4])(f: (T1,T2,T3,T4) => Z): Validation[Z] =
    Apply[Validation].map4(v1,v2,v3,v4)(f)

  /** Accumulate function for Validation[T] of arity 5 */
  def accumulate[T1,T2,T3,T4,T5,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5])(f: (T1,T2,T3,T4,T5) => Z): Validation[Z] =
    Apply[Validation].map5(v1,v2,v3,v4,v5)(f)

  /** Accumulate function for Validation[T] of arity 6 */
  def accumulate[T1,T2,T3,T4,T5,T6,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6])(f: (T1,T2,T3,T4,T5,T6) => Z): Validation[Z] =
    Apply[Validation].map6(v1,v2,v3,v4,v5,v6)(f)

  /** Accumulate function for Validation[T] of arity 7 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7])(f: (T1,T2,T3,T4,T5,T6,T7) => Z): Validation[Z] =
    Apply[Validation].map7(v1,v2,v3,v4,v5,v6,v7)(f)

  /** Accumulate function for Validation[T] of arity 8 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8])(f: (T1,T2,T3,T4,T5,T6,T7,T8) => Z): Validation[Z] =
    Apply[Validation].map8(v1,v2,v3,v4,v5,v6,v7,v8)(f)

  /** Accumulate function for Validation[T] of arity 9 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9) => Z): Validation[Z] =
    Apply[Validation].map9(v1,v2,v3,v4,v5,v6,v7,v8,v9)(f)

  /** Accumulate function for Validation[T] of arity 10 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10) => Z): Validation[Z] =
    Apply[Validation].map10(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10)(f)

  /** Accumulate function for Validation[T] of arity 11 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11) => Z): Validation[Z] =
    Apply[Validation].map11(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11)(f)

  /** Accumulate function for Validation[T] of arity 12 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12) => Z): Validation[Z] =
    Apply[Validation].map12(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12)(f)

  /** Accumulate function for Validation[T] of arity 13 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13) => Z): Validation[Z] =
    Apply[Validation].map13(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13)(f)

  /** Accumulate function for Validation[T] of arity 14 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14) => Z): Validation[Z] =
    Apply[Validation].map14(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14)(f)

  /** Accumulate function for Validation[T] of arity 15 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15) => Z): Validation[Z] =
    Apply[Validation].map15(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15)(f)

  /** Accumulate function for Validation[T] of arity 16 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16) => Z): Validation[Z] =
    Apply[Validation].map16(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16)(f)

  /** Accumulate function for Validation[T] of arity 17 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16],v17: Validation[T17])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17) => Z): Validation[Z] =
    Apply[Validation].map17(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16,v17)(f)

  /** Accumulate function for Validation[T] of arity 18 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16],v17: Validation[T17],v18: Validation[T18])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18) => Z): Validation[Z] =
    Apply[Validation].map18(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16,v17,v18)(f)

  /** Accumulate function for Validation[T] of arity 19 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16],v17: Validation[T17],v18: Validation[T18],v19: Validation[T19])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19) => Z): Validation[Z] =
    Apply[Validation].map19(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16,v17,v18,v19)(f)

  /** Accumulate function for Validation[T] of arity 20 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16],v17: Validation[T17],v18: Validation[T18],v19: Validation[T19],v20: Validation[T20])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20) => Z): Validation[Z] =
    Apply[Validation].map20(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16,v17,v18,v19,v20)(f)

  /** Accumulate function for Validation[T] of arity 21 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20,T21,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16],v17: Validation[T17],v18: Validation[T18],v19: Validation[T19],v20: Validation[T20],v21: Validation[T21])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20,T21) => Z): Validation[Z] =
    Apply[Validation].map21(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16,v17,v18,v19,v20,v21)(f)

  /** Accumulate function for Validation[T] of arity 22 */
  def accumulate[T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20,T21,T22,Z](v1: Validation[T1],v2: Validation[T2],v3: Validation[T3],v4: Validation[T4],v5: Validation[T5],v6: Validation[T6],v7: Validation[T7],v8: Validation[T8],v9: Validation[T9],v10: Validation[T10],v11: Validation[T11],v12: Validation[T12],v13: Validation[T13],v14: Validation[T14],v15: Validation[T15],v16: Validation[T16],v17: Validation[T17],v18: Validation[T18],v19: Validation[T19],v20: Validation[T20],v21: Validation[T21],v22: Validation[T22])(f: (T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20,T21,T22) => Z): Validation[Z] =
    Apply[Validation].map22(v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12,v13,v14,v15,v16,v17,v18,v19,v20,v21,v22)(f)

}
 
